modules/biobank/php/containercontroller.class.inc

```diff
@@ -0,0 +1,405 @@
+    public function validateInstance(Container $container) : void
+    {
+        $this->validateBarcode($container);
+        $this->validateTypeId($container);
+        $this->validateTemperature($container);
+        $this->validateStatusId($container);
+        $this->validateCenterId($container);
+        $this->validateParentContainerId($container);
+        $this->validateCoordinate($container);
+    }
+
+    /**
+     * Validates Container Object Barcode
+     *
+     * @param Container $container Container to be checked.
+     *
+     * @throws BadRequest if the provided Container does not meet validation requirements
+     *
+     * @return void
+     */
+    private function validateBarcode($container) : void
+    {
+        $barcode = $container->getBarcode();
+        if (is_null($barcode)) {
+            throw new \BadRequest('Barcode value must be set');
+        }
+
+        $id = $container->getId();
+        if (!isset($id)) {
+            $containerList = $this->dao->selectInstances();
+            foreach ($containerList as $c) {
+                $b = $c->getBarcode();
+                if ($b === $barcode) {
+                    throw new \BadRequest('Barcode must be unique');
+                }
+            }
+        }
+
+        //TODO: Regex check on Barcode.
+        //i.e.
+        //$specimenTypes = $this->dao->getSpecimenTypes();
+        //$regex = $specimenTypes[$specimen['typeId']]['regex'];
+        //if (isset($regex)) {
+        //    if (preg_match($regex, $container['barcode']) !== 1) {
+        //        throw new \BadRequest('Barcode is not of proper format for the
+        //                               selected specimen type');
+        //    }
+        //}
+    }
+
+    /**
+     * Validates Container Object Type ID.
+     *
+     * @param Container $container Container to be checked.
+     *
+     * @throws BadRequest if the provided Container does not meet validation requirements
+     *
+     * @return void
+     */
+    private function validateTypeId($container) : void
+    {
+        if (is_null($container->getTypeId())) {
+            throw new \BadRequest('Type value must be set');
+        }
+    }
+
+    /**
+     * Validates Container Object Temperature.
+     *
+     * @param Container $container Container to be checked.
+     *
+     * @throws BadRequest if the provided Container does not meet validation requirements
+     *
+     * @return void
+     */
+    private function validateTemperature($container) : void
+    {
+        $temperature = $container->getTemperature();
+        if (is_null($temperature)) {
+            throw new \BadRequest('Temperature value must be set.');
+        }
+        if (!(is_numeric($temperature) && is_float(floatval($temperature)))) {
+            throw new \BadRequest('Temperature must be a number.');
+        }
+        if ($temperature < -273 || $temperature > 999) {
+            throw new \BadRequest('Temperature must be between -273.15 and 1000.');
```

Comment:
I might be mistaken but should the used scale for -273.15 & 1000 be stated as well?

Reply:
I'm not exactly sure what you mean.